DenisovPavel #Lab2

Ppasha9 opened the pull request.

Commit:
Add daemon

- "first" final version of daemon

Commit:
Fix pid file mistakes

- create pid file in 'build.sh' script
- do not remove pid file in daemon code

Commit:
Fix mistakes

- use 'openlog' function
- exchange defines WARNING, INFO and ERROR by directly calling 'syslog' function
- delete unused logging

Ppasha9 committed:
Add second lab

SWAGLORD1 commented:
Не могли бы вы обойтись без включения bits/types/siginfo_t.h? Я извиняюсь, но на моем компе этот заголовочный файл отказывается подключатся, несмотря на то, что библиотека с ним вроде как установлена. Решения из интернета так же не помогли.

Ppasha9 committed:
Fix include header

- replace "bits/types/siginfo_t.h" by "csignal"

Ppasha9 commented:
Поменял bits/types/siginfo_t.h на csignal.
Теперь вроде должно работать.

Ppasha9 committed:
Add errno.h

- add errno.h header in conn.h

Ppasha9 commented:
Добавил заголовочный файл errno.h, так как я везде использую errno
Странно, что у меня спокойно собиралось и работало без него, как и у студентка Козлова
Надеюсь теперь соберется

Ppasha9 committed:
Hot fix #1

- remove unused compile parameter from makefile
- correct log messages in client and host

Ppasha9 commented:
Внес некоторые правки
Убрал ненужный параметр в makefile и подправил сообщения в клиенте и хосте

KotsMikhail commented on denisov.pavel/lab2/client/client.h:
```diff
@@ -0,0 +1,33 @@
+private:
+    const int MIN_RAND       = 1;
+    const int ALIVE_MAX_RAND = 100;
+    const int DEAD_MAX_RAND  = 50;
```
Возможно, эти переменные должны быть статическими

Ppasha9 replied:
Поправил

KotsMikhail commented on denisov.pavel/lab2/connections/src/conn_fifo.cpp:
```diff
@@ -0,0 +1,87 @@
+const char * const FIFO_PATH = "/tmp/lab2_fifo_path";
+
+bool wasCreated = false, isHost;
+int fifoID;
```
От глобальных переменных следует избавиться

Ppasha9 replied:
Поправил

KotsMikhail commented on denisov.pavel/lab2/client/client_implement.cpp:
```diff
@@ -0,0 +1,117 @@
+{
+    Message msg;
+    while (true) {
+        sem_wait(semaphore_client);
```
Ожидание семафора без таймаута

Ppasha9 replied:
Если у нас теперь хост ждет ввода числа с клавиатуры, то получается, что здесь должно быть ожидание семафора без таймаута. Иначе, если пользователь на хосте будет долго думать над новым числом, игра завершится. Разве так должно быть?

KotsMikhail replied:
Вы правы, тут действительно не должно быть таймаута, моя ошибка

KotsMikhail commented on denisov.pavel/lab2/host/host.h:
```diff
@@ -0,0 +1,62 @@
+    int curNumber;
+
+    Host ();
+    int  GetNewRandomNumber       ();
```
Дублирование кода клиента. Возможно, эту функцию лучше вынести отдельно?

Ppasha9 replied:
Убрал это у хоста, так как теперь числа вводятся с клавиатуры.
У клиента оставил, сделал статик функцией

KotsMikhail reviewed (changes requested):
После запуска и завершения хоста с именованным каналом, при повторном его запуске к нему невозможно подключить клиента, хост падает с сообщением об ошибке соединения

Ppasha9 committed:
Fix issues

- make const variables in host static
- delete global variables from connections' implementations
- now you should input new number in host every round
- add SIGINT singal support in host

Ppasha9 reviewed (commented):
Поправил все замечания.
Насчет таймаута у клиента: так и не понял зачем он нужен, ведь пользователь на хосте теперь вводит числа с клавиатуры.
Насчет проблемы с повторным запуском хоста с именованным каналом. Возможно, хост был завершен через комбинацию клавиш Ctrl + C, так что добавил обработку сигнала SIGINT в хост. Теперь должно работать, по крайней мере у меня работает корректно.

Ppasha9 committed:
Cosmetic fixes

KotsMikhail commented on denisov.pavel/lab2/client/client.h:
```diff
@@ -0,0 +1,37 @@
+
+    sem_t *semaphore_host, *semaphore_client;
+
+    explicit Client (pid_t pid);
```
Что делает ключевое слово explicit и зачем оно тут?

Ppasha9 replied:
Убрал explicit.
По сути, он не нужен. Просто я это добавил, чтобы нельзя было написать Client cl = pid, хотя у нас клиент является синглтоном, и такой ситуации просто не может быть.

KotsMikhail commented on denisov.pavel/lab2/connections/conn_interfaces/conn.h:
```diff
@@ -0,0 +1,30 @@
+private:
+    bool isHost, wasCreated;
+    int descID;
+    int *additionalArgs;
```
Не очень хорошее решение, так как теряется семантика дополнительных полей и ваш код становится очень тяжело читать.  Предлагаю хранить указатель на область памяти, в реализациях объявлять структуру и кастовать эту область памяти к ней.

Ppasha9 replied:
Переделал на void *pmem, то есть на указатель на область памяти, который в реализациях просто кастую к нужной структуре. Надеюсь теперь код более читаемый)

KotsMikhail commented on denisov.pavel/lab2/connections/conn_interfaces/conn.h:
```diff
@@ -0,0 +1,30 @@
+    bool Close ();
+
+private:
+    bool isHost, wasCreated;
```
Переменная wasCreated не используется в conn_mq. Может, это тоже дополнительный аргумент?

Ppasha9 replied:
Согласен, не заметил. Вынес переменную wasCreated в дополнительные аргументы.

KotsMikhail reviewed (changes requested):
Очередность ходов должна быть обратной. Сначала козлята делают свой ход, затем волк их ловит.
Вариант программы с сокетом никак не реагирует на ввод данных с клавиатуры после установки соединения.
Программа позволяет вводить любые числа и неадекватно реагирует на ввод строки

Ppasha9 committed:
Fix issues

- remove 'explicit' in client.h
- new order of host/client communication
- new support of additional arguments in connections' implementation
- add hack for socket connection implementation, in case when client failed and new client is attached

Ppasha9 commented:
Поменял очередность ходов. Теперь сначала ходит козленок, потом волк. Ошибку с реализацией сокетов поправил. Вроде должно работать. 
Так же добавил проверку ввода пользователя на стороне хоста. Теперь при невалидном вводе хост пишет что именно неправильно и просит попробовать еще раз.

KotsMikhail reviewed (changes requested):
Не собирается, сообщение об ошибке следующее:

./client/client_implement.cpp: In member function ‘bool Client::OpenConnection()’:
./client/client_implement.cpp:48:5: error: expected ‘;’ before ‘std’
     std::cout << "Semaphores created. Client's pid: " << getpid() << std::endl;
     ^
./client/client_implement.cpp:48:79: error: statement has no effect [-Werror=unused-value]
     std::cout << "Semaphores created. Client's pid: " << getpid() << std::endl;

Ppasha9 committed:
Hot fix

Ppasha9 commented:
Тупая ошибка.
Поправил.

SWAGLORD1 commented:
Вариант программы с сокетом зависает после установки соединения с клиентом и не реагирует на ввод с клавиатуры

Commit:
Fix waiting client section in host

Ppasha9 commented:
Проверил на Linux Mint 19 Cinnamon, и действительно хост зависает при подключении клиента. Странно что на Ubuntu 18.04 все работало прекрасно.
В любом случае, поправил ошибку.

SWAGLORD1 commented:
Засчитано
